envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py

```diff
@@ -105,29 +170,214 @@ def test_shutdown_closes_console_pipes(self):
         self.assertTrue(console.stderr.closed)
 
+    def test_ipython_util_io_globals_restored(self):
+        original_io_stdin = IPython.utils.io.stdin
+        original_io_stdout = IPython.utils.io.stdout
+        original_io_stderr = IPython.utils.io.stderr
+        original_io_raw_print = IPython.utils.io.raw_print
+        original_io_raw_print_err = IPython.utils.io.raw_print_err
+
+        self.create_and_destroy_kernel()
+
+        self.assertIs(IPython.utils.io.stdin, original_io_stdin)
+        self.assertIs(IPython.utils.io.stdout, original_io_stdout)
+        self.assertIs(IPython.utils.io.stderr, original_io_stderr)
+        self.assertIs(IPython.utils.io.raw_print, original_io_raw_print)
+        self.assertIs(
+            IPython.utils.io.raw_print_err, original_io_raw_print_err)
+
     def test_io_pub_thread_stopped(self):
+        self.create_and_destroy_kernel()
+        io_pub_threads = self.objects_of_type(ipykernel.iostream.IOPubThread)
+        for thread in io_pub_threads:
+            self.assertFalse(thread.thread.is_alive())
+
+    def test_no_threads_leaked(self):
+        threads_before = threading.active_count()
+        self.create_and_destroy_kernel()
+        threads_after = threading.active_count()
+        self.assertEqual(threads_before, threads_after)
+
+    def test_no_new_atexit_handlers(self):
+        # Caution: this is a rather fragile and indirect test. We want
+        # to know that all cleanup has happened when shutting down the
+        # kernel, with none of that cleanup deferred to atexit handlers.
+        #
+        # Since we have no direct way to get hold of the atexit handlers on
+        # Python 3, we instead use the number of referents from the
+        # atexit module as a proxy.
+        #
+        # If this test starts failing, try adding a warmup cycle. If the
+        # first call to self.create_and_destroy_kernel adds new referents,
+        # that's not a big deal. But if every call consistently adds new
+        # referents, then there's something to be fixed.
+        atexit_handlers_before = len(gc.get_referents(atexit))
+        self.create_and_destroy_kernel()
+        atexit_handlers_after = len(gc.get_referents(atexit))
+
+        self.assertEqual(atexit_handlers_before, atexit_handlers_after)
+
+    def test_zmq_sockets_closed(self):
+        # Previously, tests were leaking file descriptors linked to
+        # zmq.Socket objects. Check that all extant sockets are closed.
+        self.create_and_destroy_kernel()
+        sockets = self.objects_of_type(zmq.Socket)
+        self.assertTrue(all(socket.closed for socket in sockets))
+
+    def test_ipykernel_live_objects(self):
+        # Check that all IPython-related objects have been cleaned up
+        # as expected.
+
+        self.create_and_destroy_kernel()
+
+        # Sometimes once isn't enough ....
+        # XXX Ideally, this wouldn't be necessary at all; see if we can fix.
```

Comment:
Remove this comment. We're not going to go chasing all the cycles - there are way too many.

Reply:
Done. And really, one `gc.collect()` should be enough (there are _theoretically_ situations where you need multiple `gc.collect()` calls to clean everything up, but this is unlikely to be one of those situations); let's avoid pandering to superstition and not use multiple calls until it becomes clear that that's necessary.